src/app/devices/devices.component.ts

```diff
@@ -123,42 +125,68 @@ export class DevicesComponent implements OnInit {
 
   bulkPowerAction (action: number): void {
     const requests: Array<Observable<any>> = []
+    this.isLoading = true
     this.selection.selected.forEach(z => {
-      requests.push(this.devicesService.sendPowerAction(z.guid, action))
+      requests.push(this.devicesService.sendPowerAction(z.guid, action).pipe(
+        catchError(err => of({ err })),
+        map(i => ({
+          ReturnValueStr: i?.Body && i.Body.ReturnValueStr ? i.Body.ReturnValueStr : 'ERROR',
+          ReturnValue: i?.Body && i.Body.ReturnValue !== undefined ? i.Body.ReturnValue : -1,
+          guid: z.guid
+        }))
+      ))
     })
 
     forkJoin(requests).subscribe(result => {
-      this.snackBar.open($localize`Power action sent successfully`, undefined, SnackbarDefaults.defaultSuccess)
+      this.isLoading = false
+      this.bulkActionResponses = result
+      this.resetResponse()
+    }, err => {
+      console.log(err)
     })
   }
 
   sendPowerAction (deviceId: string, action: number): void {
     this.isLoading = true
     this.devicesService.sendPowerAction(deviceId, action).pipe(
-      catchError(err => {
-        // TODO: handle error better
-        this.snackBar.open($localize`Error sending power action`, undefined, SnackbarDefaults.defaultError)
-        return throwError(err)
-      }),
+      catchError(err => of({ err })),
       finalize(() => {
         this.isLoading = false
       })
     ).subscribe(data => {
-      this.snackBar.open($localize`Power action sent successfully`, undefined, SnackbarDefaults.defaultSuccess)
+      this.bulkActionResponses.push({
+        ReturnValueStr: data?.Body && data.Body.ReturnValueStr ? data.Body.ReturnValueStr : 'ERROR',
+        ReturnValue: data?.Body && data.Body.ReturnValue !== undefined ? data.Body.ReturnValue : -1,
+        guid: deviceId
+      })
+      this.resetResponse()
       this.devicesService.getPowerState(deviceId).pipe(delay(2000)).subscribe(z => {
         (this.devices.data.find(y => y.guid === deviceId) as any).powerstate = z.powerstate
       })
       console.log(data)
+    }, err => {
+      console.log(err)
     })
   }
 
+  resetResponse (): void {
```

Comment:
I'm not sure this is the best way to handle it. I think the errors should remain until they are tried again -- don't time them out. I need to be able to read what each one says. That's going to take more than 5 seconds or any given time. The successes can still time out though I think.